packages/flutter/lib/src/widgets/basic.dart

```diff
@@ -2369,42 +2476,138 @@ class UnconstrainedBox extends SingleChildRenderObjectWidget {
       ..clipBehavior = clipBehavior;
   }
 
   @override
-  RenderUnconstrainedBox createRenderObject(BuildContext context) => RenderUnconstrainedBox(
-    textDirection: textDirection ?? Directionality.maybeOf(context),
-    alignment: alignment,
-    constrainedAxis: constrainedAxis,
-    clipBehavior: clipBehavior,
-  );
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<AlignmentGeometry>('alignment', alignment));
+    properties.add(EnumProperty<TextDirection>('textDirection', textDirection, defaultValue: null));
+
+    final String? debugTransformLabel = _debugTransformLabel.isNotEmpty
+      ? _debugTransformLabel
+      : _knownTransforms[constraintsTransform];
+
+    if (debugTransformLabel != null) {
+      properties.add(DiagnosticsProperty<String>('constraints transform', debugTransformLabel));
+    }
+  }
+}
+
+/// A widget that imposes no constraints on its child, allowing it to render
+/// at its "natural" size.
+///
+/// This allows a child to render at the size it would render if it were alone
+/// on an infinite canvas with no constraints. This container will then attempt
+/// to adopt the same size, within the limits of its own constraints. If it ends
+/// up with a different size, it will align the child based on [alignment].
+/// If the box cannot expand enough to accommodate the entire child, the
+/// child will be clipped.
+///
+/// In debug mode, if the child overflows the container, a warning will be
+/// printed on the console, and black and yellow striped areas will appear where
+/// the overflow occurs.
+///
+/// See also:
+///
+///  * [ConstrainedBox], for a box which imposes constraints on its child.
+///  * [Align], which loosens the constraints given to the child rather than
+///    removing them entirely.
+///  * [Container], a convenience widget that combines common painting,
+///    positioning, and sizing widgets.
+///  * [OverflowBox], a widget that imposes different constraints on its child
+///    than it gets from its parent, possibly allowing the child to overflow
+///    the parent.
+class UnconstrainedBox extends ConstraintsTransformBox {
+  /// Creates a widget that imposes no constraints on its child, allowing it to
+  /// render at its "natural" size. If the child overflows the parents
+  /// constraints, a warning will be given in debug mode.
+  const UnconstrainedBox({
+    Key? key,
+    Widget? child,
+    TextDirection? textDirection,
+    AlignmentGeometry alignment = Alignment.center,
+    this.constrainedAxis,
+    Clip clipBehavior = Clip.none,
+  }) : assert(alignment != null),
+       assert(clipBehavior != null),
+       super(
+         key: key,
+         child: child,
+         textDirection: textDirection,
+         alignment: alignment,
+         constraintsTransform: _fail,
+         clipBehavior: clipBehavior,
+       );
+
+  static BoxConstraints _fail(BoxConstraints incoming) {
```

Comment:
This is kinda odd. How about defining UnconstrainedBox as a Stateless widget that builds a ConstraintsTransformBox as a child (composition over inheritance)?